tests/conftest.py

```diff
@@ -1,9 +1,172 @@
+from imitation.data.wrappers import RolloutInfoWrapper
+from imitation.policies.base import FeedForward32Policy, NormalizeFeaturesExtractor
 from imitation.util import logger
+from imitation.util.networks import RunningNorm
+
+CARTPOLE_ENV_NAME = "CartPole-v1"
+
+
+def load_or_train_ppo(
+    cache_path: str,
+    training_function: Callable[[gym.Env], PPO],
+    venv,
+) -> PPO:
+    try:
+        return PPO.load(cache_path, venv)
+    except OSError:
+        pass
+    except AssertionError:
+        pass
+    expert = training_function(venv)
+    expert.save(cache_path)
+    return expert
+
+
+def load_or_rollout_trajectories(cache_path, policy, venv) -> List[TrajectoryWithRew]:
+    try:
+        with open(cache_path, "rb") as f:
+            return pickle.load(f)
+    except OSError:
+        pass
+    except AssertionError:
+        pass
+    rollout.rollout_and_save(
+        cache_path,
+        policy,
+        venv,
+        rollout.make_sample_until(min_timesteps=2000, min_episodes=57),
+    )
+    with open(cache_path, "rb") as f:
+        return pickle.load(f)  # TODO: not re-loading the trajectory would be nicer here
+
+
+@pytest.fixture(params=[1, 4])
+def cartpole_venv(request) -> gym.Env:
+    num_envs = request.param
+    return DummyVecEnv(
+        [
+            lambda: RolloutInfoWrapper(gym.make(CARTPOLE_ENV_NAME))
+            for _ in range(num_envs)
+        ],
+    )
+
+
+def train_cartpole_expert(cartpole_venv) -> PPO:
+    policy_kwargs = dict(
+        features_extractor_class=NormalizeFeaturesExtractor,
+        features_extractor_kwargs=dict(normalize_class=RunningNorm),
+    )
+    policy = PPO(
+        policy=FeedForward32Policy,
+        policy_kwargs=policy_kwargs,
+        env=VecNormalize(cartpole_venv, norm_obs=False),
+        seed=0,
+        batch_size=64,
+        ent_coef=0.0,
+        learning_rate=0.0003,
+        n_epochs=10,
+        n_steps=64 // cartpole_venv.num_envs,
+    )
+    policy.learn(100000)
+    return policy
+
+
+@pytest.fixture
+def cartpole_expert_policy(cartpole_venv, pytestconfig) -> BasePolicy:
+    cached_expert_path = str(
+        pytestconfig.cache.makedir("experts") / CARTPOLE_ENV_NAME / "model.zip",
+    )
+    return load_or_train_ppo(
+        cached_expert_path,
+        train_cartpole_expert,
+        cartpole_venv,
+    ).policy
+
+
+@pytest.fixture
+def cartpole_expert_trajectories(
+    cartpole_expert_policy,
+    cartpole_venv,
+    pytestconfig,
+) -> List[TrajectoryWithRew]:
+    rollouts_path = str(
+        pytestconfig.cache.makedir("experts") / CARTPOLE_ENV_NAME / "rollout.pkl",
+    )
+    return load_or_rollout_trajectories(
+        rollouts_path,
+        cartpole_expert_policy,
+        cartpole_venv,
+    )
+
+
+PENDULUM_ENV_NAME = "Pendulum-v1"
+
+
+@pytest.fixture
+def pendulum_venv() -> gym.Env:
+    return DummyVecEnv([lambda: RolloutInfoWrapper(gym.make(PENDULUM_ENV_NAME))] * 8)
+
+
+def train_pendulum_expert(pendulum_venv) -> PPO:
```

Comment:
Code for this + below is very similar to CartPole, I wonder if it's worth trying to be more DRY. Although most ways I can think of to do that are sufficiently magical that I'm not sure it's worth it, at least so long as we only have two environments.

Reply:
Yes, here I decided, that the LOC and duplication we save by generalizing are not worth the added complexity and readability of the code.